### PR TITLE
Ironwood support

### PR DESCRIPTION
**Pull request by nullcopy**
Adds support for the Ironwood shielded pool (NU6.3) and the ZIP 229 v6 transaction format: parsing, compact blocks, tree states, subtree roots, mempool, and `poolTypes` filtering, plus darkside test-framework support. Uses the finalized NU6.3 consensus IDs; as with v5, `nConsensusBranchId` is read but not validated since it changes at every network upgrade.

`CompactTx.actions` is renamed to `orchardActions` — the field number is unchanged (binary wire-compatible), but the proto3 JSON key and generated client APIs change.

The first three commits are standalone fixes: darkside tree-state and subtree-roots edge cases, a `vin`/`vout` leak in `GetBlockRangeNullifiers`, and a `GetMempoolTx` panic when a transaction leaves the mempool mid-refresh.

Notes: each commit builds and passes tests independently; generated files (`walletrpc/*.pb.go`, `docs/rtd/index.html`) are regenerated per commit. No cache migration is needed — an older lightwalletd stalls at the first v6 transaction, so nothing stale is ever cached. Requires a backend with NU6.3 support (e.g. zebra 6.0.0+).

### CODE REVIEW

File: lightwallet-protocol/walletrpc/compact_formats.proto [hunk 1]
```diff
@@ -77,6 +79,12 @@ message CompactTx {
 
     // A sequence of transparent outputs being created by the transaction.
     repeated TxOut vout = 8;
+
+    // A sequence of Ironwood actions in the transaction. Ironwood reuses the
+    // compact Orchard action shape because Ironwood action fields are identical
+    // to Orchard action fields (per ZIP 229, Ironwood Action descriptions use
+    // the same OrchardAction encoding).
+    repeated CompactOrchardAction ironwoodActions = 9;
```
Comment (nullcopy) [inline lightwallet-protocol/walletrpc/compact_formats.proto]: Wouldn't it be better to group this up above next to `orchardActions`?

File: common/common.go [hunk 2]
```diff
@@ -538,13 +538,13 @@ func FilterTxPool(tx *walletrpc.CompactTx, poolTypes []walletrpc.PoolType) *wall
...
 	if slices.Contains(poolTypes, walletrpc.PoolType_ORCHARD) {
-		r.Actions = tx.Actions
+		r.OrchardActions = tx.OrchardActions
 	}
 	if len(r.Vin) > 0 ||
 		len(r.Vout) > 0 ||
 		len(r.Spends) > 0 ||
 		len(r.Outputs) > 0 ||
-		len(r.Actions) > 0 {
+		len(r.OrchardActions) > 0 {
```
Comment (nullcopy) [inline common/common.go]: What about `IronwoodActions`?

Comment (nullcopy) [reply common/common.go]: All throughout this PR

File: parser/transaction.go [hunk 3]
```diff
@@ -405,11 +411,12 @@ func (tx *Transaction) ToCompact(index int) *walletrpc.CompactTx {
...
-		Outputs:        make([]*walletrpc.CompactSaplingOutput, len(tx.shieldedOutputs)),
-		OrchardActions: make([]*walletrpc.CompactOrchardAction, len(tx.orchardActions)),
-		Vin:            make([]*walletrpc.CompactTxIn, vinLen),
-		Vout:           make([]*walletrpc.TxOut, len(tx.transparentOutputs)),
+		Spends:          make([]*walletrpc.CompactSaplingSpend, len(tx.shieldedSpends)),
+		Outputs:         make([]*walletrpc.CompactSaplingOutput, len(tx.shieldedOutputs)),
+		OrchardActions:  make([]*walletrpc.CompactOrchardAction, len(tx.orchardActions)),
+		Vin:             make([]*walletrpc.CompactTxIn, vinLen),
+		Vout:            make([]*walletrpc.TxOut, len(tx.transparentOutputs)),
+		IronwoodActions: make([]*walletrpc.CompactOrchardAction, len(tx.ironwoodActions)),
```
Comment (nullcopy) [inline parser/transaction.go]: Would be better to organize next to `OrchardActions`, no?

File: common/common.go (hunk 2 above)
Comment (nullcopy) [reply common/common.go]: Nvm, I see this is added with the ZIP229 commit, but I wonder if it makes sense for that commit to be separate. It doesn't make sense to add Ironwood fields without the v6 transaction format.

File: lightwallet-protocol/walletrpc/compact_formats.proto [hunk 4]
```diff
@@ -60,13 +61,20 @@ message CompactTx {
...
     uint32 fee = 3;
 
     repeated CompactSaplingSpend spends = 4;
     repeated CompactSaplingOutput outputs = 5;
     repeated CompactOrchardAction orchardActions = 6;
 
+    // A sequence of Ironwood actions in the transaction. Ironwood reuses the
+    // compact Orchard action shape because Ironwood action fields are identical
+    // to Orchard action fields (per ZIP 229, Ironwood Action descriptions use
+    // the same OrchardAction encoding).
```
Comment (nullcopy) [inline lightwallet-protocol/walletrpc/compact_formats.proto]: This comment is superfluous. Remove it.

**Review by nullcopy — COMMENTED** [review]
Self-ack 80c845e6bb26b9830ebea8e2fc0d5c809e8bd943. I left one nit-pick comment, but not worth blocking the PR.

**Review by pacu — CHANGES_REQUESTED** [review]
At a glance I don't see that this repo is using https://github.com/zcash/lightwallet-protocol/releases/tag/v0.5.0  via `git tree` 

changes to the implementation must not only be conformant to the canonical files but they should also reference them

**Comment by nullcopy** [comment]
Good catch @pacu! I didn't realize the lightwallet-protocol was vendored as a subtree at all :upside_down_face:. Fixed.

**Review by pacu — APPROVED** [review]
utACK

File: common/common.go [hunk 5]
```diff
@@ -334,7 +343,7 @@ func getBlockFromRPC(height int) (*walletrpc.CompactBlock, error) {
 	}
 	// Fetch the block using the verbose option ("1") because it provides
 	// both the list of txids, which we're not yet able to compute for
-	// Orchard (V5) transactions, and the block hash (block ID), which
+	// v5 and later transactions, and the block hash (block ID), which
```
Comment (daira) [inline common/common.go:346]: Why not? Just not implemented?

File: common/common.go [hunk 6]
```diff
@@ -540,11 +550,15 @@ func FilterTxPool(tx *walletrpc.CompactTx, poolTypes []walletrpc.PoolType) *wall
...
 	}
+	if slices.Contains(poolTypes, walletrpc.PoolType_IRONWOOD) {
+		r.IronwoodActions = tx.IronwoodActions
+	}
 	if len(r.Vin) > 0 ||
 		len(r.Vout) > 0 ||
 		len(r.Spends) > 0 ||
 		len(r.Outputs) > 0 ||
-		len(r.Actions) > 0 {
+		len(r.Actions) > 0 ||
```
Comment (daira) [inline common/common.go:560]: I would have renamed this to `OrchardActions`.

File: lightwallet-protocol/walletrpc/service.proto [hunk 7]
```diff
@@ -181,11 +182,13 @@ message TreeState {
     uint32 time = 4;        // Unix epoch time when the block was mined
     string saplingTree = 5; // sapling commitment tree state
     string orchardTree = 6; // orchard commitment tree state
+    string ironwoodTree = 7; // ironwood commitment tree state
 }
 
 enum ShieldedProtocol {
```
Comment (daira) [inline lightwallet-protocol/walletrpc/service.proto:188]: Ironwood is not a shielded protocol; it's a shielded pool. Conflating pools with protocols is something I very strictly and thoroughly avoided when writing the NU6.3 specs. 

I know this is actually used as an argument to `GetSubtreeRoots`, which should have been a pool type. The misnaming should at least be documented.

```suggestion
// Despite the name, this type is used to indicate a shielded pool in the argument to `GetSubtreeRoots`.
// That is its only use, and because it is misnamed, other uses should not be added.
// Note that Ironwood is a shielded pool only, not a shielded protocol. 
enum ShoeldedProtocol {
```

File: lightwallet-protocol/walletrpc/service.proto [hunk 8]
```diff
@@ -312,7 +317,7 @@ service CompactTxStreamer {
     rpc GetLatestTreeState(Empty) returns (TreeState) {}
 
     // Returns a stream of information about roots of subtrees of the note commitment tree
-    // for the specified shielded protocol (Sapling or Orchard).
+    // for the specified shielded protocol (Sapling, Orchard, or Ironwood).
```
Comment (daira) [inline lightwallet-protocol/walletrpc/service.proto:320]:
```suggestion
    // for the specified shielded *pool* (Sapling, Orchard, or Ironwood). The `ShieldedProtocol`
    // type is misnamed and actually refers to a shielded pool.
```

File: lightwallet-protocol/CHANGELOG.md [hunk 9]
```diff
@@ -5,7 +5,26 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
...
 
-## Unreleased
+## [v0.5.0] - 2026-06-30
+
+### Added
+- `compact_formats.ChainMetadata` has added field `ironwoodCommitmentTreeSize`.
+- `compact_formats.CompactTx` has added field `ironwoodActions`, which may be
+  used to represent Ironwood action data.
+- `service.PoolType` has added value `IRONWOOD`.
+- `service.ShieldedProtocol` has added value `ironwood`.
```
Comment (daira) [inline lightwallet-protocol/CHANGELOG.md:15]:
```suggestion
- `service.ShieldedProtocol` has added value `ironwood`. Note that `ShieldedProtocol` is
  a misnomer; this type is only used to refer to a shielded *pool* in the argument to the
  `GetSubtreeRoots` RPC.
```

File: frontend/service.go [hunk 10]
```diff
@@ -837,6 +850,7 @@ func (s *lwdStreamer) GetSubtreeRoots(arg *walletrpc.GetSubtreeRootsArg, resp wa
 	switch arg.ShieldedProtocol {
 	case walletrpc.ShieldedProtocol_sapling:
 	case walletrpc.ShieldedProtocol_orchard:
+	case walletrpc.ShieldedProtocol_ironwood:
 		break
 	default:
 		return errors.New("unrecognized shielded protocol")
```
Comment (daira) [inline frontend/service.go:856]:
```suggestion
		return errors.New("unrecognized ShieldedProtocol (i.e. pool)")
```

File: parser/transaction.go [hunk 11]
```diff
@@ -511,12 +521,60 @@ func (tx *Transaction) parseV5(data []byte) ([]byte, error) {
...
+// inputs/outputs, Sapling bundle, Orchard bundle, then the new Ironwood
+// bundle. The Ironwood bundle reuses the Orchard Action encoding per ZIP 229.
+func (tx *Transaction) parseV6(data []byte) ([]byte, error) {
+	s := bytestring.String(data)
+	var err error
+	if !s.ReadUint32(&tx.consensusBranchID) {
+		return nil, errors.New("could not read nConsensusBranchId")
 	}
-	if tx.nVersionGroupID != 0x26A7270A {
+	if tx.nVersionGroupID != NU6_3_VERSION_GROUP_ID {
```
Comment (daira) [inline parser/transaction.go:570]: Version group IDs are associated with transaction formats, not with NUs.
```suggestion
	if tx.nVersionGroupID != ZIP229_VERSION_GROUP_ID {
```

File: parser/transaction.go [hunk 12]
```diff
@@ -511,12 +521,60 @@ func (tx *Transaction) parseV5(data []byte) ([]byte, error) {
...
 		// This shouldn't be possible
-		return nil, fmt.Errorf("version group ID %d must be 0x26A7270A", tx.nVersionGroupID)
+		return nil, fmt.Errorf("version group ID 0x%08X must be 0x%08X",
+			tx.nVersionGroupID, NU6_3_VERSION_GROUP_ID)
```
Comment (daira) [inline parser/transaction.go:573]:
```suggestion
			tx.nVersionGroupID, ZIP229_VERSION_GROUP_ID)
```

File: parser/transaction.go [hunk 13]
```diff
@@ -511,12 +521,60 @@ func (tx *Transaction) parseV5(data []byte) ([]byte, error) {
...
 	}
+	// Like parseV5, do not validate nConsensusBranchId: it identifies the
+	// consensus epoch the transaction is mined in, not the epoch that
+	// introduced the v6 format, so it changes at every network upgrade.
```
Comment (daira) [inline parser/transaction.go:577]: This comment is misplaced; it should be before the read of `nConsensusBranchId` (the first `ReadUint32` call in this function).

File: parser/transaction.go [hunk 14]
```diff
@@ -528,6 +586,34 @@ func (tx *Transaction) parseV5(data []byte) ([]byte, error) {
...
+	}
+	tx.orchardActions = orchardActions
+
+	var ironwoodActions []action
+	s, ironwoodActions, err = parseOrchardActionShapeBundle([]byte(s), "Ironwood")
+	if err != nil {
+		return nil, err
+	}
+	tx.ironwoodActions = ironwoodActions
+
```
Comment (daira) [inline parser/transaction.go:607]: As for v5, this ignores trailing data, allowing invalid transactions to be parsed.

File: parser/transaction.go [hunk 15]
```diff
@@ -511,12 +521,60 @@ func (tx *Transaction) parseV5(data []byte) ([]byte, error) {
...
+		return nil, err
+	}
+
+	var orchardActions []action
+	s, orchardActions, err = parseOrchardActionShapeBundle([]byte(s), "Orchard")
+	if err != nil {
+		return nil, err
+	}
+	tx.orchardActions = orchardActions
+
```
Comment (daira) [inline parser/transaction.go:553]: This ignores trailing data, allowing invalid transactions to be parsed.

File: parser/transaction.go [hunk 16]
```diff
@@ -575,59 +661,73 @@ func (tx *Transaction) parseV5(data []byte) ([]byte, error) {
...
 const SAPLING_TX_VERSION uint32 = 4
 const ZIP225_TX_VERSION uint32 = 5
+const ZIP229_TX_VERSION uint32 = 6
 
 const OVERWINTER_VERSION_GROUP_ID uint32 = 0x03C48270
 const SAPLING_VERSION_GROUP_ID uint32 = 0x892F2085
 const ZIP225_VERSION_GROUP_ID uint32 = 0x26A7270A
 
+// NU6.3 version group ID and consensus branch ID, per ZIP 258.
+const NU6_3_VERSION_GROUP_ID uint32 = 0xD884B698
```
Comment (daira) [inline parser/transaction.go:728]:
```suggestion
// Version group IDs are defined in §7.1.2 Transaction Consensus Rules.
const OVERWINTER_VERSION_GROUP_ID uint32 = 0x03C48270
const SAPLING_VERSION_GROUP_ID uint32 = 0x892F2085
const ZIP225_VERSION_GROUP_ID uint32 = 0x26A7270A
const ZIP229_VERSION_GROUP_ID uint32 = 0xD884B698

// NU6.3 consensus branch ID, per ZIP 258.
```

File: parser/transaction.go [hunk 17]
```diff
@@ -646,6 +746,12 @@ func (tx *Transaction) isZip225V5() bool {
 		tx.version == ZIP225_TX_VERSION
 }
 
+func (tx *Transaction) isZip229V6() bool {
+	return tx.fOverwintered &&
+		tx.nVersionGroupID == NU6_3_VERSION_GROUP_ID &&
```
Comment (daira) [inline parser/transaction.go:751]:
```suggestion
		tx.nVersionGroupID == ZIP229_VERSION_GROUP_ID &&
```

File: parser/transaction_test.go [hunk 18]
```diff
@@ -113,3 +114,139 @@ func TestV5TransactionParser(t *testing.T) {
 		}
 	}
 }
+
+func TestZip229V6TransactionParser(t *testing.T) {
+	// header: fOverwintered | version 6
+	// nVersionGroupId: 0xD884B698 (NU6.3)
```
Comment (daira) [inline parser/transaction_test.go:120]:
```suggestion
	// nVersionGroupId: 0xD884B698 (ZIP 229)
```

File: parser/transaction_test.go [hunk 19]
```diff
@@ -113,3 +114,139 @@ func TestV5TransactionParser(t *testing.T) {
...
+	if err != nil {
+		t.Fatalf("%v", err)
+	}
+	if len(rest) != 0 {
+		t.Fatalf("Test did not consume entire buffer, %d remaining", len(rest))
+	}
+	if tx.version != ZIP229_TX_VERSION {
+		t.Fatal("version miscompare")
+	}
+	if tx.nVersionGroupID != NU6_3_VERSION_GROUP_ID {
```
Comment (daira) [inline parser/transaction_test.go:143]:
```suggestion
	if tx.nVersionGroupID != ZIP229_VERSION_GROUP_ID {
```

File: parser/transaction_test.go [hunk 20]
```diff
@@ -113,3 +114,139 @@ func TestV5TransactionParser(t *testing.T) {
...
+	if tx.consensusBranchID != 0x4C89A1F3 {
+		t.Fatal("consensusBranchID miscompare")
+	}
+}
+
+func TestZip229V6TransactionParserKeepsIronwoodBundle(t *testing.T) {
+	var raw bytes.Buffer
+	raw.Write([]byte{
+		0x06, 0x00, 0x00, 0x80, // fOverwintered | version 6
+		0x98, 0xb6, 0x84, 0xd8, // version group ID (NU6.3)
```
Comment (daira) [inline parser/transaction_test.go:197]:
```suggestion
		0x98, 0xb6, 0x84, 0xd8, // version group ID (ZIP 229)
```

**Review by daira — COMMENTED** [review]
Post-hoc ACK with comments. I will submit a follow-up PR.

File: common/common.go (hunk 6 above)
Comment (nullcopy) [reply common/common.go:560]: I left this same comment myself, but the lightwallet-protocol definition did not adopt that naming convention, so I'm staying consistent with that.
